language/move-prover/src/prelude.bpl

```diff
@@ -963,17 +993,56 @@ procedure {:inline 1} $U64Util_u64_to_bytes(val: Value) returns (res: Value)  {
 // ==================================================================================
 // Native hash
 
-// TODO: implement the below methods
-
-procedure {:inline 1} $Hash_sha2_256(val: Value) returns (res: Value)  {
-    res := DefaultValue;
-    assert false; // $Hash_sha2_256 not implemented
-}
-
-procedure {:inline 1} $Hash_sha3_256(val: Value) returns (res: Value)  {
-    res := DefaultValue;
-    assert false; // $Hash_sha3_256 not implemented
-}
+// Hash is modeled as an otherwise uninterpreted injection.
+// In truth, it is not an injection since the domain has greater cardinality
+// (arbitrary length vectors) than the co-domain (vectors of length 32).  But it is
+// common to assume in code there are no hash collisions in practice.  Fortunately,
+// Boogie is not smart enough to recognized that there is an inconsistency.
+// FIXME: If we had a reliable extensional theory of arrays, and if we could use ==
+// instead of IsEqual, we might be able to avoid so many quantified formulas by
+// using a sha2_inverse function in the ensures conditions of Hash_sha2_256 to
+// assert that sha2/3 are injections without using global quantified axioms.
+
+function $sha2(val: Value) : Value;
+
+// This says that sha2 respects isEquals (this would be automatic if we had an
+// extensional theory of arrays and used ==, which has the substitution property
```

Comment:
This  comment is inaccurate.  We do have an extensional theory of arrays which we could use with ==.  What we do not have and need is an extensional theory of partial arrays (or finite-length arrays).

Reply:
You're right.  The axiomatized theory we are USING is not extensional, which caused some grief here.  In the interests of landing this, I'm going to do bors-libra now and fix this comment in the next PR.